File: evaluations/pandapower/tests/expressiveness/test_a3_dcopf.py
"""
Test A-3: Solve DC OPF with gen costs and line flow limits

Dimension: expressiveness
Network: TINY (IEEE 39-bus New England)
Pass condition: Converges. Optimal dispatch and LMPs/shadow prices extractable
    from solution. With differentiated costs and 70% derating, at least 2 branches
    have non-zero shadow prices (binding flow constraints). Report max LMP spread
    across buses.
Tool: pandapower 3.4.0
"""

from __future__ import annotations

import json
import sys
import time
import traceback
from pathlib import Path

import numpy as np
import pandas as pd

sys.path.insert(0, str(Path(__file__).resolve().parents[3] / "shared"))

from matpower_loader import load_pandapower

# Differentiated cost curves from data/timeseries/case39/README.md
COST_BY_TECH = {
    "hydro": {"cp1": 5.0, "cp2": 0.005},
    "nuclear": {"cp1": 10.0, "cp2": 0.010},
    "coal_large": {"cp1": 25.0, "cp2": 0.025},
    "gas_CC": {"cp1": 40.0, "cp2": 0.040},
}

BRANCH_DERATING = 0.70


def run(
    network_file: str = "data/networks/case39.m",
    timeseries_dir: str | None = "data/timeseries/case39",
) -> dict:
    """Execute DC OPF test and return structured results."""
    results: dict = {
        "status": "fail",
        "wall_clock_seconds": 0.0,
        "details": {},
        "errors": [],
        "workarounds": [],
    }

    start = time.perf_counter()
    try:
        import pandapower as pp

        # 1. Load network
        net = load_pandapower(network_file)
        results["details"]["bus_count"] = len(net.bus)
        results["details"]["line_count"] = len(net.line)
        results["details"]["trafo_count"] = len(net.trafo)
        results["details"]["gen_count"] = len(net.gen)
        results["details"]["ext_grid_count"] = len(net.ext_grid)

        # 2. Load differentiated costs from timeseries dir
        if timeseries_dir is None:
            results["errors"].append("timeseries_dir is required for A-3 TINY")
            return results

        ts_dir = Path(timeseries_dir)
        gen_params = pd.read_csv(ts_dir / "gen_temporal_params.csv")
        results["details"]["gen_params_loaded"] = len(gen_params)

        # 3. Apply differentiated costs to generators and ext_grid
        # pandapower OPF requires cost functions via create_poly_cost
        # First, set controllable=True and P limits on generators
        for idx in net.gen.index:
            net.gen.at[idx, "controllable"] = True
            net.gen.at[idx, "min_p_mw"] = 0.0
            # max_p_mw is already set from MATPOWER data

        # ext_grid is the slack bus — also needs cost and controllable for OPF
        for idx in net.ext_grid.index:
            net.ext_grid.at[idx, "controllable"] = True
            # Set reasonable P limits for ext_grid
            net.ext_grid.at[idx, "min_p_mw"] = -9999.0
            net.ext_grid.at[idx, "max_p_mw"] = 9999.0

        # Set bus voltage limits (required for OPF even in DC mode)
        net.bus["min_vm_pu"] = 0.9
        net.bus["max_vm_pu"] = 1.1

        # Clear existing cost functions (from_mpc may have created them)
        net.poly_cost.drop(net.poly_cost.index, inplace=True)
        if hasattr(net, "pwl_cost"):
            net.pwl_cost.drop(net.pwl_cost.index, inplace=True)

        # Apply cost functions using gen_temporal_params tech_class_key
        # gen_params is indexed by gen_index (0-based matching MATPOWER order)
        # pandapower gen table may differ in indexing from MATPOWER after conversion
        # The ext_grid (bus 39 in case39) is gen_index=9 in MATPOWER

        # Map MATPOWER gen indices to pandapower elements
        # In pandapower from_mpc, one generator becomes ext_grid (the slack),
        # the rest become gen elements
        for _, row in gen_params.iterrows():
            gen_idx_mat = int(row["gen_index"])
            tech = row["tech_class_key"]
            costs = COST_BY_TECH.get(tech, COST_BY_TECH["gas_CC"])
            # gen_temporal_params uses 1-indexed MATPOWER bus IDs;
            # pandapower from_mpc uses 0-indexed buses
            bus_id_pp = int(row["bus_id"]) - 1

            # Find the pandapower element for this generator
            # Check ext_grid first
            ext_match = net.ext_grid[net.ext_grid["bus"] == bus_id_pp]
            gen_match = net.gen[net.gen["bus"] == bus_id_pp]

            if len(ext_match) > 0:
                eidx = ext_match.index[0]
                pp.create_poly_cost(
                    net,
                    element=eidx,
                    et="ext_grid",
                    cp1_eur_per_mw=costs["cp1"],
                    cp2_eur_per_mw2=costs["cp2"],
                    cp0_eur=0.0,
                )
            elif len(gen_match) > 0:
                gidx = gen_match.index[0]
                pp.create_poly_cost(
                    net,
                    element=gidx,
                    et="gen",
                    cp1_eur_per_mw=costs["cp1"],
                    cp2_eur_per_mw2=costs["cp2"],
                    cp0_eur=0.0,
                )
            else:
                results["errors"].append(
                    f"Could not find pandapower element for MATPOWER gen_index={gen_idx_mat} "
                    f"at pp bus {bus_id_pp} (MATPOWER bus {int(row['bus_id'])})"
                )

        results["details"]["cost_functions_created"] = len(net.poly_cost)

        # 4. Apply branch derating (70%)
        # Derate lines
        net.line["max_loading_percent"] = 100.0  # Ensure the column exists
        original_max_i_ka = net.line["max_i_ka"].copy()
        net.line["max_i_ka"] = original_max_i_ka * BRANCH_DERATING

        # Derate transformers
        if len(net.trafo) > 0:
            net.trafo["max_loading_percent"] = 100.0

        results["details"]["branch_derating"] = BRANCH_DERATING

        # 5. Solve DC OPF
        solve_start = time.perf_counter()
        pp.rundcopp(net)
        solve_time = time.perf_counter() - solve_start
        results["details"]["solve_seconds"] = solve_time

        # 6. Check convergence
        opf_converged = net.OPF_converged
        results["details"]["opf_converged"] = opf_converged
        if not opf_converged:
            results["errors"].append("DC OPF did not converge")
            return results

        # 7. Extract dispatch results
        gen_dispatch = net.res_gen["p_mw"].to_dict()
        ext_grid_dispatch = net.res_ext_grid["p_mw"].to_dict()
        results["details"]["gen_dispatch_mw"] = gen_dispatch
        results["details"]["ext_grid_dispatch_mw"] = ext_grid_dispatch
        results["details"]["total_generation_mw"] = float(
            net.res_gen["p_mw"].sum() + net.res_ext_grid["p_mw"].sum()
        )
        results["details"]["total_load_mw"] = float(net.load["p_mw"].sum())

        # 8. Extract LMPs / shadow prices
        # pandapower's PYPOWER OPF stores bus shadow prices in net._ppc
        lmp_extracted = False
        lmp_source = None
        bus_lmps = {}

        # Try extracting from internal PYPOWER result
        try:
            ppc = net._ppc
            # PYPOWER stores bus marginal prices in bus column 13 (LAM_P)
            if ppc is not None and "bus" in ppc:
                bus_data = ppc["bus"]
                # Column indices: LAM_P=13, LAM_Q=14, MU_VMAX=15, MU_VMIN=16
                if bus_data.shape[1] > 13:
                    lam_p = bus_data[:, 13]
                    results["details"]["lam_p_raw"] = lam_p.tolist()
                    if np.any(np.abs(lam_p) > 1e-10):
                        lmp_extracted = True
                        lmp_source = "net._ppc['bus'][:, 13] (LAM_P)"
                        # Map internal bus indices back to pandapower bus indices
                        # _ppc uses internal indexing; need pd2ppc lookups
                        try:
                            bus_lookup = net._pd2ppc_lookups.get("bus", None)
                            if bus_lookup is not None:
                                for pp_bus_idx in net.bus.index:
                                    internal_idx = bus_lookup[pp_bus_idx]
                                    bus_lmps[int(pp_bus_idx)] = float(lam_p[internal_idx])
                        except Exception:
                            # Fallback: assume sequential mapping
                            for i, val in enumerate(lam_p):
                                bus_lmps[i] = float(val)
        except (AttributeError, KeyError, IndexError) as e:
            results["details"]["lmp_extraction_error"] = str(e)

        # Also check res_bus for any LMP-like columns
        res_bus_cols = list(net.res_bus.columns)
        results["details"]["res_bus_columns"] = res_bus_cols
        lmp_cols = [c for c in res_bus_cols if "lam" in c.lower() or "lmp" in c.lower()]
        results["details"]["lmp_columns_in_res_bus"] = lmp_cols

        results["details"]["lmp_extracted"] = lmp_extracted
        results["details"]["lmp_source"] = lmp_source
        results["details"]["bus_lmps"] = bus_lmps

        if bus_lmps:
            lmp_values = list(bus_lmps.values())
            max_lmp = max(lmp_values)
            min_lmp = min(lmp_values)
            lmp_spread = max_lmp - min_lmp
            results["details"]["max_lmp"] = max_lmp
            results["details"]["min_lmp"] = min_lmp
            results["details"]["lmp_spread"] = lmp_spread
        else:
            results["details"]["lmp_spread"] = None

        # 9. Extract branch shadow prices / binding constraints
        binding_branches = 0
        branch_shadow_prices = {}

        try:
            ppc = net._ppc
            if ppc is not None and "branch" in ppc:
                branch_data = ppc["branch"]
                # MU_SF=13, MU_ST=14 (shadow prices on from/to flow limits)
                if branch_data.shape[1] > 14:
                    mu_sf = branch_data[:, 13]
                    mu_st = branch_data[:, 14]
                    for i in range(len(mu_sf)):
                        mu = max(abs(mu_sf[i]), abs(mu_st[i]))
                        if mu > 1e-6:
                            binding_branches += 1
                            branch_shadow_prices[i] = {
                                "mu_sf": float(mu_sf[i]),
                                "mu_st": float(mu_st[i]),
                            }
        except (AttributeError, KeyError, IndexError) as e:
            results["details"]["shadow_price_extraction_error"] = str(e)

        results["details"]["binding_branches"] = binding_branches
        results["details"]["branch_shadow_prices"] = branch_shadow_prices

        # 10. Check line loading
        line_loading = net.res_line["loading_percent"]
        results["details"]["max_line_loading_pct"] = float(line_loading.max())
        results["details"]["lines_above_95pct_loading"] = int((line_loading > 95.0).sum())

        # 11. Objective value
        try:
            ppc = net._ppc
            if ppc is not None and "f" in ppc:
                results["details"]["objective_value"] = float(ppc["f"])
        except (AttributeError, KeyError):
            pass

        results["details"]["output_format"] = "pandas.DataFrame + net._ppc internals"

        # 12. Check pass conditions
        if not lmp_extracted:
            results["workarounds"].append(
                "LMPs extracted from internal PYPOWER structure (net._ppc['bus'][:, 13]) "
                "rather than a public API. This is a fragile workaround that accesses "
                "private attributes."
            )

        if binding_branches < 2:
            results["errors"].append(
                f"Only {binding_branches} branches have non-zero shadow prices "
                f"(need >= 2 with 70% derating and differentiated costs)"
            )
            # Still report results but don't pass
        else:
            results["status"] = "pass" if lmp_extracted else "qualified_pass"

    except Exception as e:
        results["errors"].append(f"{type(e).__name__}: {e}")
        results["details"]["traceback"] = traceback.format_exc()
    finally:
        results["wall_clock_seconds"] = time.perf_counter() - start

    return results


if __name__ == "__main__":
    result = run()
    print(json.dumps(result, indent=2, default=str))

File: evaluations/matpower/tests/expressiveness/test_a3_dcopf.m
%% Test A-3: Solve DC OPF with differentiated gen costs and 70% branch derating on TINY
%%
%% Dimension: expressiveness
%% Network: TINY (IEEE 39-bus New England)
%% Pass condition: Converges. Optimal dispatch and LMPs/shadow prices extractable.
%%   With differentiated costs and 70% derating, at least 2 branches have non-zero
%%   shadow prices (binding flow constraints). Report max LMP spread across buses.
%% Tool: MATPOWER 8.1
%% Solver: HiGHS

%% Setup MATPOWER paths
mp_root = fullfile(fileparts(mfilename('fullpath')), '..', '..', 'matpower8.1');
addpath(fullfile(mp_root, 'lib'));
addpath(fullfile(mp_root, 'data'));
addpath(fullfile(mp_root, 'mips', 'lib'));
addpath(fullfile(mp_root, 'mp-opt-model', 'lib'));
addpath(fullfile(mp_root, 'mptest', 'lib'));

network_file = '/workspace/data/networks/case39.m';
timeseries_dir = '/workspace/data/timeseries/case39';

result_status = 'fail';
errors = {};
workarounds = {};
solver_used = 'unknown';

try
    %% Load case
    mpc = loadcase(network_file);
    define_constants;

    %% Load differentiated costs from gen_temporal_params.csv
    gen_params_file = fullfile(timeseries_dir, 'gen_temporal_params.csv');
    fid = fopen(gen_params_file, 'r');
    if fid == -1
        error('Cannot open %s', gen_params_file);
    end
    header = fgetl(fid);
    gen_data = {};
    while ~feof(fid)
        line = fgetl(fid);
        if ischar(line) && ~isempty(line)
            gen_data{end + 1} = line;
        end
    end
    fclose(fid);

    % Parse tech_class_key (column 5, 1-indexed)
    tech_keys = {};
    for i = 1:length(gen_data)
        parts = strsplit(gen_data{i}, ',');
        tech_keys{i} = strtrim(parts{5});
    end

    % Cost map: tech_class_key -> [c1 ($/MWh), c2 ($/MW^2h)]
    % cost = c2*Pg^2 + c1*Pg + c0
    ngen = size(mpc.gen, 1);

    % Ensure gencost has enough columns for quadratic (7 columns)
    new_gencost = zeros(ngen, 7);
    for i = 1:ngen
        key = tech_keys{i};
        switch key
            case 'hydro'
                c1 = 5.0;
                c2 = 0.005;
            case 'nuclear'
                c1 = 10.0;
                c2 = 0.010;
            case 'coal_large'
                c1 = 25.0;
                c2 = 0.025;
            case 'gas_CC'
                c1 = 40.0;
                c2 = 0.040;
            otherwise
                c1 = 40.0;
                c2 = 0.040;
        end
        % Polynomial type=2, startup=0, shutdown=0, ncost=3, c2, c1, c0
        new_gencost(i, :) = [2 0 0 3 c2 c1 0];
    end
    mpc.gencost = new_gencost;

    %% Apply 70% branch derating
    mpc.branch(:, RATE_A) = mpc.branch(:, RATE_A) * 0.70;
    mpc.branch(:, RATE_B) = mpc.branch(:, RATE_B) * 0.70;
    mpc.branch(:, RATE_C) = mpc.branch(:, RATE_C) * 0.70;

    %% Check if HiGHS is available
    highs_available = have_feature('highs');
    fprintf('HiGHS available: %d\n', highs_available);

    %% Configure solver
    if highs_available
        mpopt = mpoption('verbose', 0, 'out.all', 0, 'opf.dc.solver', 'HIGHS');
        solver_used = 'HiGHS';
    else
        fprintf('HiGHS not available, using MIPS (built-in)\n');
        mpopt = mpoption('verbose', 0, 'out.all', 0, 'opf.dc.solver', 'MIPS');
        solver_used = 'MIPS';
        workarounds{end + 1} = 'HiGHS solver not available; using MIPS (built-in)';
    end

    %% Solve DC OPF
    tic;
    results = rundcopf(mpc, mpopt);
    solve_time = toc;

    if ~results.success
        errors{end + 1} = sprintf('DC OPF did not converge with %s', solver_used);

        %% Try fallback to GLPK if primary solver failed
        if ~strcmp(solver_used, 'MIPS')
            fprintf('Primary solver failed, trying MIPS...\n');
            mpopt_fb = mpoption('verbose', 0, 'out.all', 0, 'opf.dc.solver', 'MIPS');
            tic;
            results = rundcopf(mpc, mpopt_fb);
            solve_time = toc;
            if results.success
                solver_used = 'MIPS (fallback)';
                workarounds{end + 1} = sprintf('Primary solver failed; fell back to MIPS');
                errors = {};
            end
        end
    end

    if results.success
        %% Extract results
        bus_count = size(results.bus, 1);
        branch_count = size(results.branch, 1);

        % LMPs
        lmps = results.bus(:, LAM_P);
        lmp_max = max(lmps);
        lmp_min = min(lmps);
        lmp_spread = lmp_max - lmp_min;

        % Shadow prices on branch flow limits
        mu_sf = results.branch(:, MU_SF);
        mu_st = results.branch(:, MU_ST);
        binding_branches = (abs(mu_sf) > 1e-4) | (abs(mu_st) > 1e-4);
        binding_count = sum(binding_branches);

        % Generator dispatch
        gen_pg = results.gen(:, PG);
        gen_buses = results.gen(:, GEN_BUS);

        % Objective value
        obj_value = results.f;

        %% Peak memory
        [~, mem_out] = system('grep VmHWM /proc/self/status');
        peak_kb = sscanf(mem_out, 'VmHWM: %f');
        if ~isempty(peak_kb)
            peak_memory_mb = peak_kb / 1024;
        else
            peak_memory_mb = -1;
        end

        %% Print results
        fprintf('\n=== Test A-3: DC OPF Results ===\n');
        fprintf('Wall clock: %.4f s\n', solve_time);
        fprintf('Solver: %s\n', solver_used);
        fprintf('Objective value: $%.2f\n', obj_value);
        fprintf('Total generation: %.2f MW\n', sum(gen_pg));
        fprintf('Total load: %.2f MW\n', sum(results.bus(:, PD)));
        fprintf('LMP range: [%.4f, %.4f] $/MWh, spread: %.4f\n', lmp_min, lmp_max, lmp_spread);
        fprintf('LMP mean: %.4f $/MWh\n', mean(lmps));
        fprintf('Binding branches: %d\n', binding_count);
        fprintf('Branch derating: 70%%\n');
        fprintf('Peak memory: %.1f MB\n', peak_memory_mb);

        % Print binding branch details
        binding_idx = find(binding_branches);
        if ~isempty(binding_idx)
            fprintf('\nBinding branch details:\n');
            fprintf('  From -> To  | mu_sf     | mu_st     | Flow (MW)\n');
            for i = 1:length(binding_idx)
                idx = binding_idx(i);
                fprintf('  %3d -> %3d | %9.4f | %9.4f | %9.2f\n', ...
                        results.branch(idx, F_BUS), results.branch(idx, T_BUS), ...
                        mu_sf(idx), mu_st(idx), results.branch(idx, PF));
            end
        end

        % Print dispatch
        fprintf('\nGenerator dispatch:\n');
        fprintf('  Bus | Tech     | Dispatch (MW) | Pmax (MW)\n');
        for i = 1:ngen
            fprintf('  %3d | %-8s | %12.2f  | %9.2f\n', ...
                    gen_buses(i), tech_keys{i}, gen_pg(i), results.gen(i, PMAX));
        end

        % Print LMPs per bus (all buses)
        fprintf('\nBus LMPs:\n');
        fprintf('  Bus | LMP ($/MWh) | Load (MW)\n');
        for i = 1:bus_count
            fprintf('  %3d | %11.4f | %9.2f\n', ...
                    results.bus(i, BUS_I), lmps(i), results.bus(i, PD));
        end

        %% Pass condition check
        if binding_count >= 2 && lmp_spread > 0
            result_status = 'pass';
        else
            if binding_count < 2
                errors{end + 1} = sprintf('Only %d binding branches (need >= 2)', binding_count);
            end
            if lmp_spread <= 0
                errors{end + 1} = 'No LMP spread (all buses have same price)';
            end
        end
    end

catch e
    errors{end + 1} = e.message;
end

fprintf('\nStatus: %s\n', result_status);
if ~isempty(errors)
    fprintf('Errors: %s\n', strjoin(errors, '; '));
end
if ~isempty(workarounds)
    fprintf('Workarounds: %s\n', strjoin(workarounds, '; '));
end
